### PR TITLE
Docs: Show source of corresponding args in source block

### LINKED ISSUE #20745: [Bug]: Source snippets clobber each other when you render a story more than once
### Describe the bug

When you render a story twice on the page (as in autodocs), it is keyed in the `SourceProvider` by id. This means when you change the primary story's args, the new args show up in the new secondary one.

### To Reproduce

1. Go to http://localhost:6006/?path=/docs/example-button--docs (any sandbox)
2. Show code in the Primary story in the `Stories` block
3. Change args. The story will not change but the code will.

### PR DESCRIPTION
**Pull request by tmeasday**
Closes #20745

## What I did

- Emit args with `SNIPPET_RENDERED` event
- Store code in source container keyed on args hash
- Read code based on args except if `__forceInitialArgs` is passed to `Source` (in which case read on `initialArgs`).
- Pass `__forceInitialArgs` through.

## How to test

See stories & E2E tests. Try changing args on docs page with source open and check only the primary story changes.

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/renderers/react/src/docs/jsxDecorator.tsx [hunk 1]
```diff
@@ -190,7 +190,8 @@ export const jsxDecorator = (
 
   useEffect(() => {
     if (!skip) {
-      channel.emit(SNIPPET_RENDERED, (context || {}).id, jsx);
+      const { id, args } = context || {};
+      channel.emit(SNIPPET_RENDERED, id, jsx, false, args);
```
Comment (JReinhold) [inline code/renderers/react/src/docs/jsxDecorator.tsx:194]: given the amount of arguments to this event, we should convert them to a single object argument instead.

**Review by JReinhold — CHANGES_REQUESTED** [review]
Code looks good.

Still concerned about this though https://github.com/storybookjs/storybook/issues/20745#issuecomment-1415214819

This would require all renderers to change right, even the community ones we don't control?

File: code/renderers/react/src/docs/jsxDecorator.tsx (hunk 1 above)
Comment (tmeasday) [reply code/renderers/react/src/docs/jsxDecorator.tsx:194]: I was aiming for backwards compatibility here (I guess we could do that even with changing it to an object based argument).

**Comment by tmeasday** [comment]
@JReinhold note the fallback to "unknown args" when we don't receive args details on the channel message.

File: code/ui/blocks/src/blocks/SourceContainer.tsx [hunk 2]
```diff
@@ -21,24 +29,51 @@ export interface SourceContextProps {
...
 export const SourceContext: Context<SourceContextProps> = createContext({ sources: {} });
 
+type SnippetRenderedEvent = {
+  id: StoryId;
+  source: string;
+  args?: Args;
+  format?: SyntaxHighlighterFormatTypes;
+};
+
+export const unknownArgsHash = '--unknown--';
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/SourceContainer.tsx:39]: nit: this looks like a constant.

```suggestion
export const UNKNOWN_ARGS_HASH = '--unknown--';
```

**Review by JReinhold — APPROVED** [review]
LGTM! Only a minor nit.

**Comment by tmeasday** [comment]
@shilman as discussed, the E2E test I added here which inspects the arg value in the snippet is failing in various frameworks that don't appear to be rendering the snippet (should contain something like `label="Basic"`) correctly:

[Angular](https://app.circleci.com/pipelines/github/storybookjs/storybook/40986/workflows/051fca6d-9369-406b-b074-8d3dc2a99f60/jobs/482673/parallel-runs/0?filterBy=FAILED): 
```
<storybook-framework-button
      label=\"false\"
      (onClick)=\"onClick($event)\"
    ></storybook-framework-button>
```

[Lit:](https://app.circleci.com/pipelines/github/storybookjs/storybook/40986/workflows/051fca6d-9369-406b-b074-8d3dc2a99f60/jobs/482673/parallel-runs/9?filterBy=FAILED)
```
<sb-button></sb-button>
```

[Next:](https://app.circleci.com/pipelines/github/storybookjs/storybook/40986/workflows/051fca6d-9369-406b-b074-8d3dc2a99f60/jobs/482673/parallel-runs/11?filterBy=FAILED)  - known issue: https://github.com/storybookjs/storybook/issues/20356

```
<head_manager_provider_default>
  <No Display Name />
</head_manager_provider_default>"
```

**Pull request merged**